test/integration/xss_protection_test.rb

```diff
@@ -0,0 +1,18 @@
+require 'test_helper'
+
+class ActionDispatch::IntegrationTest
+  include Capybara::DSL
+
+  test "should escape html correctly" do
+    visit('/bookmarks/new')
+    fill_in('Title', with: "Yamaha Saxophones")
+    fill_in('Link', with: "http://usa.yamaha.com/products/musical-instruments/winds/sax/")
+    fill_in('Description', with: "usa homepage for yamaha saxophones &`<'>\",!@$()[]{}=+")
+    click_button 'Submit'
+    bookmark = Bookmark.find_by_id("980190963")
+    assert_equal "Yamaha&nbsp;Saxophones", bookmark.title
+    assert_equal "http://usa.yamaha.com/products/musical-instruments/winds/sax/", bookmark.link
+    assert_equal "usa&nbsp;homepage&nbsp;for&nbsp;yamaha&nbsp;saxophones&nbsp;&amp;`&lt;'&gt;\&quot;,!@$()[]{}=+",
+                 bookmark.description
+  end
```

Comment:
Sanitizing happens in different places to protect against different types of attacks. [SQL injection](https://en.wikipedia.org/wiki/SQL_injection) attacks, for example, are defended against in ActiveRecord by escaping input characters to db queries that could be parsed as SQL instead of input values.

It wouldn't be appropriate for ActiveRecord to clean inputs to protect against XSS attacks, though, because that's a concern related to View layer (by View I do not specifically mean templates, but rather the logic that is involved with presenting are rendering data).

Tests like `assert_equal "Yamaha&nbsp;Saxophones", bookmark.title` will always fail, then, because those characters aren't getting escaped there. However, both ERB and HAML automatically escape all strings passed in at render time. That means that to assert things are getting escaped properly to prevent XSS attacks, you've got to use Capybara to assert that injected javascript won't actually run:

```ruby
require 'test_helper'

def assert_no_console_logs
  assert_equal 0, page.driver.console_messages.size
end

class ActionDispatch::IntegrationTest
  include Capybara::DSL

  test "should escape html correctly" do
    visit('/bookmarks/new')
    fill_in('Title', with: "Yamaha Saxophones Bookmark Title<script>console.log('uh oh');</script>")
    fill_in('Link', with: "http://usa.yamaha.com/products/musical-instruments/winds/sax/")
    fill_in('Description', with: "usa homepage for yamaha saxophones &`<'>\",!@$()[]{}=+")
    click_button 'Submit'

    assert page.has_content? "Yamaha Saxophones Bookmark Title"
    assert_no_console_logs
  end
end
```

To make the above test work you'll need to run Capybara using the [`webkit`](https://github.com/thoughtbot/capybara-webkit) driver so that it can run Javascript. To see the test fail, change line 7 of your `show.html.haml` file to `%td.title!= @bookmark.title`. The `!` prevents HAML from html-escaping the input, which will cause the injected console log script to get parsed as HTML and run. Then we can ask the webkit driver for any console logs that happened, if you want to see them you can throw `pp page.driver.console_messages` in the test.